### DISCUSSION TITLE
ARToolKitPlus ported to Typescript

### DISCUSSION THREAD

**Discussion by kalwalt** [AR-js-org/AR.js#491] [Ideas]
[ARToolKitPlus ](https://github.com/paroj/artoolkitplus) as explained in the [README.md](https://github.com/paroj/artoolkitplus#artoolkitplus)  `is an extended version of ARToolKit's vision code that adds features, but breaks compatibility due to a new class-based API.` Maybe it worth to develop a web version and i started to port to Javascript and Typescript in this my repository [artoolkitplus-ts](https://github.com/kalwalt/artoolkitplus-ts). Not all the code is ported yet, and i have to say this project in an infancy stage, but there are some rough example to test if you want to try. 
@nicolocarpignoli @nickw1 it could be that this project could be an alternative approach to **ARToolKit**, maybe it could be in a future integrated into AR.js? In any case i will move the project to the **[AR-js-org](https://github.com/AR-js-org)** 
I will not have so much time in this period of the year ( but when i have...??) but i hope to move further soon. 
There are some differences between **artoolkit5-js** (ARToolkit) and **artoolkitplus-ts** (ARToolKitPlus):

|             | artoolkit5-js | artoolkitplus-ts |
|-------------|---------------|------------------|
| source code |   C           | C++              |
| camera calibration data | binary file | text file (OpenCV.js standard) |
| image data  | allocating pointers through EM_ASM | passing directly data through emscripten::val |
| detecting markers | marker data is passed through global variable artoolkit in the `markerInfo` object | data passed internally no gloabl variable required ! |
| pose |  | use rpp: Robust Planar Pose a more efficient pose estimation algorithm |

I will improve this table of comparison as i will have more time, i hope that you will find useful.

## Documentation

For a general documentation: You can read this [PDF](https://arbook.icg.tugraz.at/schmalstieg/Schmalstieg_114.pdf)
For a comparison between **ARToolKit** and **ARToolKItPlus** read this [article](https://answers.launchpad.net/artoolkitplus/+faq/775)

**Comment by nickw1** [comment]
@kalwalt with the caveat that I'm not so familiar with the code base on the marker / NFT side, this looks good.

**Reply by kalwalt** [reply]
> @kalwalt with the caveat that I'm not so familiar with the code base on the marker / NFT side, this looks good.

No worries @nickw1 i appreciate in any case, Merry Xmas 🎄 🎉 🥇

**Comment by nickw1** [comment]
@kalwalt thanks, Merry Christmas too!

**Comment by kalwalt** [comment]
I added Typescript support to ARToolKitPlus, there is also a basic video example, within you can point a marker and you got a message in the console with the id number and the matrix position. I will do a more elaborate example in a future.

**Comment by kalwalt** [comment]
Now [artoolkitplus-ts](https://github.com/AR-js-org/artoolkitplus-ts) is part of the AR.js org and i published the first npm package.
